app/src/main/java/emu/skyline/EmulationActivity.kt

```diff
@@ -342,6 +393,79 @@ class EmulationActivity : AppCompatActivity(), SurfaceHolder.Callback, View.OnTo
         }
     }
 
+    private fun getPictureInPictureBuilder() : PictureInPictureParams.Builder {
+        val pictureInPictureParamsBuilder = PictureInPictureParams.Builder()
+
+        val pictureInPictureActions : MutableList<RemoteAction> = mutableListOf()
+        val pendingFlags = PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
+
+        val pauseIcon = Icon.createWithResource(this, R.drawable.ic_pause)
+        val pausePendingIntent = PendingIntent.getBroadcast(this, R.drawable.ic_pause, Intent(intentActionPause), pendingFlags)
+        val pauseRemoteAction = RemoteAction(pauseIcon, getString(R.string.pause), getString(R.string.pause_emulator), pausePendingIntent)
+        pictureInPictureActions.add(pauseRemoteAction)
+
+        if (!emulationSettings.isAudioOutputDisabled) {
+            val muteIcon = Icon.createWithResource(this, R.drawable.ic_volume_mute)
+            val mutePendingIntent = PendingIntent.getBroadcast(this, R.drawable.ic_volume_mute, Intent(intentActionMute), pendingFlags)
+            val muteRemoteAction = RemoteAction(muteIcon, getString(R.string.mute), getString(R.string.disable_audio_output), mutePendingIntent)
+            pictureInPictureActions.add(muteRemoteAction)
+        }
+
+        pictureInPictureParamsBuilder.setActions(pictureInPictureActions)
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S)
+            pictureInPictureParamsBuilder.setAutoEnterEnabled(true)
+
+        setPictureInPictureParams(pictureInPictureParamsBuilder.build())
+
+        return pictureInPictureParamsBuilder
+    }
+
+    override fun onPictureInPictureModeChanged(isInPictureInPictureMode: Boolean, newConfig: Configuration) {
+        super.onPictureInPictureModeChanged(isInPictureInPictureMode, newConfig)
+        if (isInPictureInPictureMode) {
+            pictureInPictureReceiver = object : BroadcastReceiver() {
+                override fun onReceive(context : Context?, intent : Intent) {
+                    if (intent.action == intentActionPause)
+                        pauseEmulator()
+                    else if (intent.action == intentActionMute)
+                        changeAudioStatus(false)
+                }
+            }
+
+            IntentFilter().apply {
+                addAction(intentActionPause)
+                if (!emulationSettings.isAudioOutputDisabled)
+                    addAction(intentActionMute)
+            }.also {
+                registerReceiver(pictureInPictureReceiver, it)
+            }
+
+            binding.onScreenControllerView.isGone = true
+            binding.onScreenControllerToggle.isGone = true
+            binding.onScreenPauseToggle.isGone = true
+        } else {
+            try {
+                if (this::pictureInPictureReceiver.isInitialized)
```

Comment:
IMO having `pictureInPictureReceiver` a lateinit var on which you check `isInitialized` is a bad use of lateinit. What I'd suggest doing is declare the receiver like the following, right above the `onPictureInPictureModeChanged` method, outside of it:
```kt
private val pictureInPictureReceiver = object : BroadcastReceiver() {
    override fun onReceive(context : Context, intent : Intent) {
        if (intent.action == intentActionPause)
            pauseEmulator()
        else if (intent.action == intentActionMute)
            changeAudioStatus(false)
    }
}
```

Reply:
lateinit in general is a bad use of lateinit. The only purpose it really has is to not deal with ? and !!